File: apps/backend/agent-api/api/routers/quiz.py
from fastapi import APIRouter, HTTPException

from api.services import quiz_service
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/spaces", tags=["quiz"])


@router.post("/{space_slug}/topics/{topic_slug}/quiz")
def generate_quiz(space_slug: str, topic_slug: str):
    try:
        questions = quiz_service.generate_topic_quiz(space_slug, topic_slug)
        return {"questions": questions}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{space_slug}/topics/{topic_slug}/quiz")
def get_quiz(space_slug: str, topic_slug: str):
    try:
        quiz_data = quiz_service.get_topic_quiz(space_slug, topic_slug)
        return quiz_data
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{space_slug}/topics/{topic_slug}/quiz")
def delete_quiz(space_slug: str, topic_slug: str):
    try:
        quiz_service.delete_topic_quiz(space_slug, topic_slug)
        return {"message": "Quiz deleted"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: apps/frontend/web/src/api/quiz.ts
const API_BASE = 'http://localhost:8000'

export interface QuizChoice {
  id: string
  text: string
}

export interface QuizQuestion {
  id: number
  question: string
  choices: QuizChoice[]
  correctAnswer: string
}

export interface QuizResponse {
  questions: QuizQuestion[]
}

async function fetchApi<T>(url: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${url}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers,
    },
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Unknown error' }))
    throw new Error(error.detail || 'API error')
  }

  return response.json()
}

export async function generateQuiz(spaceSlug: string, topicSlug: string): Promise<QuizResponse> {
  return fetchApi<QuizResponse>(`/api/spaces/${spaceSlug}/topics/${topicSlug}/quiz`, {
    method: 'POST',
  })
}

export async function getQuiz(spaceSlug: string, topicSlug: string): Promise<QuizResponse> {
  return fetchApi<QuizResponse>(`/api/spaces/${spaceSlug}/topics/${topicSlug}/quiz`)
}

export async function deleteQuiz(spaceSlug: string, topicSlug: string): Promise<{ message: string }> {
  return fetchApi<{ message: string }>(`/api/spaces/${spaceSlug}/topics/${topicSlug}/quiz`, {
    method: 'DELETE',
  })
}
